src/transformers/models/whisper/generation_whisper.py

```diff
@@ -0,0 +1,1436 @@
+                scores_is_logprobs=num_beams > 1,
+            )
+            logits_processor = (
+                [no_speech_detector] if logits_processor is None else [no_speech_detector] + logits_processor
+            )
+            no_speech_detector.set_model(self)
+
+        if is_shortform and generation_config.forced_decoder_ids is not None:
+            forced_tokens_proc = ForceTokensLogitsProcessor(generation_config.forced_decoder_ids)
+            # TODO(Patrick): It's important that the `forced_tokens_proc` processor is appended after
+            # the suppress_tokens processor or else it might happen that all token logits are suppressed to -inf
+            # which would lead to unexpected behavior
+            # The better approach here is to NOT make use of the `forced_tokens_proc` for Whisper and instead
+            # initialize all of them as `decoder_input_ids`.
+            logits_processor = (
+                [forced_tokens_proc] if logits_processor is None else logits_processor + [forced_tokens_proc]
+            )
+            generation_config.forced_decoder_ids = None
+
+        return logits_processor
+
+    @staticmethod
+    def _maybe_reduce_batch(input_features, seek, max_frames, cur_bsz, batch_idx_map):
+        prev_bsz = cur_bsz
+        new_batch_idx_map = []
+        for i in range(prev_bsz):
+            prev_i = batch_idx_map[i]
+            if seek[prev_i] >= max_frames[prev_i]:
+                cut_index = i + (cur_bsz - prev_bsz)
+                cur_bsz -= 1
+                input_features = torch.cat([input_features[:cut_index], input_features[cut_index + 1 :]], dim=0)
+            else:
+                # cut out index that goes away
+                new_batch_idx_map.append(prev_i)
+
+        return input_features, cur_bsz, new_batch_idx_map
+
+    @staticmethod
+    def _get_input_segment(input_features, seek, seek_num_frames, num_segment_frames, cur_bsz, batch_idx_map):
+        segment_input = []
+        for i in range(cur_bsz):
+            prev_i = batch_idx_map[i]
+            segment_input_slice = input_features[i : i + 1, :, seek[prev_i] : seek[prev_i] + seek_num_frames[prev_i]]
+
+            if segment_input_slice.shape[-1] < num_segment_frames:
+                # pad to 3000 if necessary
+                segment_input_slice = F.pad(
+                    segment_input_slice, pad=(0, num_segment_frames - segment_input_slice.shape[-1])
+                )
+
+            segment_input.append(segment_input_slice)
+
+        segment_input = torch.cat(segment_input, dim=0)
+
+        return segment_input
+
+    # TODO(Patrick) - remove prev_start_of_text
+    @staticmethod
+    def _prepare_decoder_input_ids(
+        cur_bsz,
+        init_tokens,
+        current_segments,
+        batch_idx_map,
+        do_condition_on_prev_tokens,
+        generation_config,
+        config,
+        device,
+        kwargs,
+        prev_start_of_text,
```

Comment:
Once https://github.com/huggingface/transformers/pull/27658/files#r1445152333 is done, `prev_start_of_text` should be removed here as an input